Install istio latest release/master 1.7-alpha via istioctl got runtime error

May-05
Following is the error log:

```
istio-1.7-alpha.8dfdb00c9ef491c5df220780c96802efdc483d40/bin/istioctl
+ install_istioctl /home/prow/go/src/istio.io/tools/perf/istio-install/tmp/istio-1.7-alpha.8dfdb00c9ef491c5df220780c96802efdc483d40 -f istioctl_profiles/default.yaml
+ release=/home/prow/go/src/istio.io/tools/perf/istio-install/tmp/istio-1.7-alpha.8dfdb00c9ef491c5df220780c96802efdc483d40
+ shift
+ /home/prow/go/src/istio.io/tools/perf/istio-install/tmp/istio-1.7-alpha.8dfdb00c9ef491c5df220780c96802efdc483d40/bin/istioctl manifest apply --skip-confirmation --wait -d /home/prow/go/src/istio.io/tools/perf/istio-install/tmp/istio-1.7-alpha.8dfdb00c9ef491c5df220780c96802efdc483d40/manifests -f istioctl_profiles/default.yaml
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2278321]

goroutine 1 [running]:
istio.io/istio/operator/pkg/apis/istio/v1alpha1/validation.validateFeatures(0xc0002467e0, 0xc000b91a40, 0x0, 0x0, 0x0)
	istio.io/istio/operator/pkg/apis/istio/v1alpha1/validation/validation.go:111 +0x51
istio.io/istio/operator/pkg/apis/istio/v1alpha1/validation.ValidateConfig(0x2a7f600, 0xc0001fa030, 0xc000b91a40, 0xb5e, 0xc0001d85a0, 0x97, 0x2f57400, 0xc00055f600)
	istio.io/istio/operator/pkg/apis/istio/v1alpha1/validation/validation.go:45 +0x201
istio.io/istio/operator/cmd/mesh.GenerateConfig(0xc000abbbe0, 0x1, 0x1, 0xc0001d85a0, 0x97, 0xc00036c000, 0xc00055f600, 0x2f574a0, 0xc0008923c0, 0x0, ...)
	istio.io/istio/operator/cmd/mesh/profile-common.go:63 +0x187
istio.io/istio/operator/cmd/mesh.ApplyManifests(0xc0009416d0, 0x1, 0x1, 0xc000abbbe0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
	istio.io/istio/operator/cmd/mesh/manifest-apply.go:173 +0x196
istio.io/istio/operator/cmd/mesh.runApplyCmd(0xc0008c9080, 0xc000129cc8, 0xc000d10300, 0xc00006e630, 0x0, 0x0)
	istio.io/istio/operator/cmd/mesh/manifest-apply.go:149 +0x279
istio.io/istio/operator/cmd/mesh.manifestApplyCmd.func1(0xc0008c9080, 0xc0008aecc0, 0x0, 0x6, 0x0, 0x0)
	istio.io/istio/operator/cmd/mesh/manifest-apply.go:101 +0x46
github.com/spf13/cobra.(*Command).execute(0xc0008c9080, 0xc0008aec60, 0x6, 0x6, 0xc0008c9080, 0xc0008aec60)
	github.com/spf13/cobra@v1.0.0/command.go:842 +0x453
github.com/spf13/cobra.(*Command).ExecuteC(0xc0002d6000, 0x8, 0x8, 0xc0002d6000)
	github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
	istio.io/istio/istioctl/cmd/istioctl/main.go:29 +0x76
```

--- a/operator/cmd/mesh/manifest-generate_test.go
+++ b/operator/cmd/mesh/manifest-generate_test.go
@@ -429,6 +429,27 @@ func TestBareSpec(t *testing.T) {
 	}
 }
 
+func TestBareValues(t *testing.T) {
+	testDataDir = filepath.Join(operatorRootDir, "cmd/mesh/testdata/manifest-generate")
+	inPathBase := filepath.Join(testDataDir, "input/bare_values.yaml")
+	// As long as the generate doesn't panic, we pass it.  bare_values.yaml doesn't
+	// overlay well because JSON doesn't handle null values, and our charts
+	// don't expect values to be blown away.
+	_, _ = runManifestGenerate([]string{inPathBase}, "", liveCharts)
+}
+
+func TestBogusControlPlaneSec(t *testing.T) {
+	testDataDir = filepath.Join(operatorRootDir, "cmd/mesh/testdata/manifest-generate")
+	inPathBase := filepath.Join(testDataDir, "input/bogus_cps.yaml")
+	// As long as the generate doesn't panic, we pass it.  bare_values.yaml doesn't
+	// overlay well because JSON doesn't handle null values, and our charts
+	// don't expect values to be blown away.
+	_, err := runManifestGenerate([]string{inPathBase}, "", liveCharts)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
 func TestInstallPackagePath(t *testing.T) {
 	testDataDir = filepath.Join(operatorRootDir, "cmd/mesh/testdata/manifest-generate")
 	serverDir, err := ioutil.TempDir(os.TempDir(), "istio-test-server-*")

--- a/operator/cmd/mesh/manifest-generate_test.go
+++ b/operator/cmd/mesh/manifest-generate_test.go
@@ -441,9 +441,6 @@ func TestBareValues(t *testing.T) {
 func TestBogusControlPlaneSec(t *testing.T) {
 	testDataDir = filepath.Join(operatorRootDir, "cmd/mesh/testdata/manifest-generate")
 	inPathBase := filepath.Join(testDataDir, "input/bogus_cps.yaml")
-	// As long as the generate doesn't panic, we pass it.  bare_values.yaml doesn't
-	// overlay well because JSON doesn't handle null values, and our charts
-	// don't expect values to be blown away.
 	_, err := runManifestGenerate([]string{inPathBase}, "", liveCharts)
 	if err != nil {
 		t.Fatal(err)

--- a/operator/cmd/mesh/testdata/manifest-generate/input/bare_values.yaml
+++ b/operator/cmd/mesh/testdata/manifest-generate/input/bare_values.yaml
@@ -1,0 +1,6 @@
+apiVersion: install.istio.io/v1alpha1
+kind: IstioOperator
+metadata:
+  namespace: istio-system
+spec:
+  values:

--- a/operator/cmd/mesh/testdata/manifest-generate/input/bogus_cps.yaml
+++ b/operator/cmd/mesh/testdata/manifest-generate/input/bogus_cps.yaml
@@ -1,0 +1,10 @@
+apiVersion: install.istio.io/v1alpha1
+kind: IstioOperator
+metadata:
+  namespace: istio-system
+spec:
+  values:
+    global:
+      controlPlaneSecurityEnabled:
+      mtls:
+        auto: true

--- a/operator/pkg/apis/istio/v1alpha1/validation/validation.go
+++ b/operator/pkg/apis/istio/v1alpha1/validation/validation.go
@@ -108,7 +108,9 @@ func validateFeatures(values *valuesv1alpha1.Values, _ *v1alpha1.IstioOperatorSp
 	if m == nil {
 		return nil
 	}
-	if m.GetAuto().Value && !g.GetControlPlaneSecurityEnabled().Value {
+	cpse := g.GetControlPlaneSecurityEnabled()
+	cpsEnabled := cpse != nil && cpse.Value
+	if m.GetAuto().Value && !cpsEnabled {
 		return []error{fmt.Errorf("security: auto mtls is enabled, but control plane security is not enabled")}
 	}
 	return nil

--- a/operator/pkg/apis/istio/v1alpha1/validation/validation.go
+++ b/operator/pkg/apis/istio/v1alpha1/validation/validation.go
@@ -108,9 +108,7 @@ func validateFeatures(values *valuesv1alpha1.Values, _ *v1alpha1.IstioOperatorSp
 	if m == nil {
 		return nil
 	}
-	cpse := g.GetControlPlaneSecurityEnabled()
-	cpsEnabled := cpse != nil && cpse.Value
-	if m.GetAuto().Value && !cpsEnabled {
+	if m.GetAuto().Value && !g.GetControlPlaneSecurityEnabled().GetValue() {
 		return []error{fmt.Errorf("security: auto mtls is enabled, but control plane security is not enabled")}
 	}
 	return nil

--- a/operator/pkg/tpath/tpath.go
+++ b/operator/pkg/tpath/tpath.go
@@ -198,6 +198,10 @@ func getPathContext(nc *PathContext, fullPath, remainPath util.Path, createMissi
 				}
 			}
 		}
+		if reflect.ValueOf(ncNode).IsNil() {
+			ncNode = make(map[string]interface{})
+			nc.Node = ncNode
+		}
 		if m, ok := ncNode.(map[string]interface{}); ok {
 			nn, ok = m[pe]
 			if !ok {

